impl_core/src/test/java/io/opencensus/implcore/tags/TagContextImplTest.java

```diff
@@ -38,55 +38,108 @@
+  private static final TagKey K1 = TagKey.create("k1");
+  private static final TagKey K2 = TagKey.create("k2");
 
   private static final TagValue V1 = TagValue.create("v1");
   private static final TagValue V2 = TagValue.create("v2");
 
   @Rule public final ExpectedException thrown = ExpectedException.none();
 
   @Test
-  public void testSet() {
-    TagContext tags = tagger.emptyBuilder().put(KS1, V1).build();
-    assertThat(tagContextToList(tagger.toBuilder(tags).put(KS1, V2).build()))
-        .containsExactly(Tag.create(KS1, V2));
-    assertThat(tagContextToList(tagger.toBuilder(tags).put(KS2, V2).build()))
-        .containsExactly(Tag.create(KS1, V1), Tag.create(KS2, V2));
+  public void getTags_empty() {
+    TagContextImpl tags = new TagContextImpl(ImmutableMap.<TagKey, TagValue>of());
+    assertThat(tags.getTags()).isEmpty();
+  }
+
+  @Test
+  public void getTags_nonEmpty() {
+    TagContextImpl tags = new TagContextImpl(ImmutableMap.of(K1, V1, K2, V2));
+    assertThat(tags.getTags()).containsExactly(K1, V1, K2, V2);
+  }
+
+  @Test
+  public void put_newKey() {
+    TagContext tags = new TagContextImpl(ImmutableMap.of(K1, V1));
+    assertThat(((TagContextImpl) tagger.toBuilder(tags).put(K2, V2).build()).getTags())
+        .containsExactly(K1, V1, K2, V2);
+  }
+
+  @Test
+  public void put_existingKey() {
+    TagContext tags = new TagContextImpl(ImmutableMap.of(K1, V1));
+    assertThat(((TagContextImpl) tagger.toBuilder(tags).put(K1, V2).build()).getTags())
+        .containsExactly(K1, V2);
+  }
+
+  @Test
+  public void put_nullKey() {
+    TagContext tags = new TagContextImpl(ImmutableMap.of(K1, V1));
+    TagContextBuilder builder = tagger.toBuilder(tags);
+    thrown.expect(NullPointerException.class);
+    thrown.expectMessage("key");
+    builder.put(null, V2);
+  }
+
+  @Test
+  public void put_nullValue() {
+    TagContext tags = new TagContextImpl(ImmutableMap.of(K1, V1));
+    TagContextBuilder builder = tagger.toBuilder(tags);
+    thrown.expect(NullPointerException.class);
+    thrown.expectMessage("value");
+    builder.put(K2, null);
+  }
+
+  @Test
+  public void remove_existingKey() {
+    TagContext tags = new TagContextImpl(ImmutableMap.of(K1, V1, K2, V2));
+    assertThat(((TagContextImpl) tagger.toBuilder(tags).remove(K1).build()).getTags())
+        .containsExactly(K2, V2);
+  }
+
+  @Test
+  public void remove_differentKey() {
+    TagContext tags = new TagContextImpl(ImmutableMap.of(K1, V1));
+    assertThat(((TagContextImpl) tagger.toBuilder(tags).remove(K2).build()).getTags())
+        .containsExactly(K1, V1);
   }
 
   @Test
-  public void testClear() {
-    TagContext tags = tagger.emptyBuilder().put(KS1, V1).build();
-    assertThat(tagContextToList(tagger.toBuilder(tags).remove(KS1).build())).isEmpty();
-    assertThat(tagContextToList(tagger.toBuilder(tags).remove(KS2).build()))
-        .containsExactly(Tag.create(KS1, V1));
+  public void remove_nullValue() {
```

Comment:
should `remove_nullKey` be better since we are testing against the key?

Reply:
Thanks, I made a PR: #910